### PR TITLE
Synchronise components

### PR DESCRIPTION
**Pull request by tobiasgradl**

### CODE REVIEW

File: src/main/java/de/cronn/jira/sync/config/JiraProjectSync.java [hunk 1]
```diff
@@ -23,10 +23,15 @@
 	private Map<String, TransitionConfig> transitions = new LinkedHashMap<>();
 	private Map<String, String> versionMapping = new LinkedHashMap<>();
 	private Set<String> versionsToIgnore = new LinkedHashSet<>();
+	private Map<String, String> componentMapping = new LinkedHashMap<>();
+	private Set<String> componentsToIgnore = new LinkedHashSet<>();
 	private Set<String> skipUpdateInTargetWhenStatusIn = new LinkedHashSet<>();
 	private boolean copyCommentsToTarget = false;
 	private Map<String, Map<String, String>> fieldValueMappings = new LinkedHashMap<>();
 
+	public JiraProjectSync() {
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/config/JiraProjectSync.java]: shouldn’t be necessary

File: src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java [hunk 2]
```diff
@@ -79,6 +80,12 @@ public void setVersions(Set<JiraVersion> versions) {
 		this.versions = versions;
 	}
 
+	@JsonIgnore
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java]: shouldn’t be necessary for a `with…` method

File: src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java [hunk 3]
```diff
@@ -95,9 +102,19 @@ public JiraFieldsUpdate withFixVersions(Set<JiraVersion> jiraVersions) {
...
+	public Set<JiraComponent> getComponents() {
+		return components;
+	}
+
+	@Override
+	public void setComponents(Set<JiraComponent> components) {
+		this.components = components;
+	}
+
 	@JsonIgnore
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java]: we can also remove this `@JsonIgnore`

File: src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+			return true;
+		}
+
+		if (!(o instanceof JiraNamedResource)) {
+			return false;
+		}
+
+		JiraNamedResource that = (JiraNamedResource) o;
+
+		return new EqualsBuilder()
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java:50]: please use plain `Objects.equals(…)` as it was implemented before.

File: src/main/java/de/cronn/jira/sync/domain/JiraRemoteLink.java [hunk 5]
```diff
@@ -6,7 +6,7 @@
 import org.apache.commons.lang3.builder.ToStringBuilder;
 import org.apache.commons.lang3.builder.ToStringStyle;
 
-public class JiraRemoteLink extends JiraIdResource {
+public class JiraRemoteLink {
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/domain/JiraRemoteLink.java]: Why is this change good/required? How is it related to the component change?

File: src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		}
+
+
+		List<T> targetResources = targetResourceSupplier.get();
+
+		Set<T> mappedResources = new LinkedHashSet<>();
+		for (T resourceToMap : resourcesToMap) {
+			T mappedResource = map(resourceToMap, resourceNamesToIgnore, targetResources, nameMapping);
+			if (mappedResource != null) {
+				log.trace("Mapping resource: {}  -->  {}", resourceToMap, mappedResource);
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java]: two spaces

File: src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+		String mappedResourceName = nameMapping.get(resourceName);
+		if (mappedResourceName == null) {
+			log.warn("No mapping defined for '{}'", resourceToMap);
+			return null;
+		}
+
+		T mappedResource = targetResources.stream()
+			.filter(resource -> Objects.equals(resource.getName(), mappedResourceName))
+			.findFirst()
```
Comment (bwaldvogel) [inline src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java:72]: what happens if we would have two matches?

File: src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java (hunk 4 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java:50]: This kind of `equals` was previously implemented in
* `JiraIssueStatus`
* `JiraResolution`
* `JiraVersion`

For `JiraIssueStatus`, for example, it was implemented in [0674aef5](https://github.com/cronn-de/jira-sync/commit/0674aef5). Do you remember why? And should I keep the implementation for the three classes above?

File: src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java (hunk 7 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java:72]: We would take the first one. :-) (The algorithm originates from `DefaultVersionMapper`.)

Leaving that unnoticed wouldn't be good, of course. I propose to log a warning and return `null` in this case. OK?

File: src/main/java/de/cronn/jira/sync/config/JiraProjectSync.java (hunk 1 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/config/JiraProjectSync.java]: Removed.

File: src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java (hunk 2 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java]: Removed.

File: src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java (hunk 3 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/domain/JiraFieldsUpdate.java]: Removed.

File: src/main/java/de/cronn/jira/sync/domain/JiraRemoteLink.java (hunk 5 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/domain/JiraRemoteLink.java]: It doesn't seem to be required. I must have changed it accidentally.

I reverted the change.

File: src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java (hunk 6 above)
Comment (tobiasgradl) [reply src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java]: Fixed.

File: src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java (hunk 7 above)
Comment (bwaldvogel) [reply src/main/java/de/cronn/jira/sync/mapping/DefaultNamedResourceMapper.java:72]: ah right, I thought we already have something like `StreamUtils.toSingleElement()` in this project but we don’t have it yet. Could be a follow-up PR.

**Comment by bwaldvogel** [comment]
Review completed. Feel free to rebase and and then merge on master.

File: src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java (hunk 4 above)
Comment (bwaldvogel) [reply src/main/java/de/cronn/jira/sync/domain/JiraNamedResource.java:50]: To be discussed and changed as follow-up.

**Pull request merged**